File: packages/relief_contracts/python/relief_contracts/household_snapshot.py
"""HouseholdSnapshotV1 (Section 10).

Owner: Plan Two. Version: 1.0.0. The complete, self-contained input to every
forecast provider — the model team must never query Plan Two's production
database directly. Python twin of ``src/household_snapshot.ts``.
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, field_validator

from .financial_event import FinancialEventV1
from .shared import (
    AccountV1,
    ConsumerConstitutionV1,
    ObligationV1,
    ProviderCapabilityV1,
    validate_contract_version,
)


class HouseholdSnapshotV1(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contract_version: str
    snapshot_id: str
    household_id: str
    generated_at: datetime
    timezone: str
    currency: str
    accounts: list[AccountV1]
    obligations: list[ObligationV1]
    recent_events: list[FinancialEventV1]
    known_future_events: list[FinancialEventV1]
    consumer_constitution: ConsumerConstitutionV1
    provider_capabilities: list[ProviderCapabilityV1]

    @field_validator("contract_version")
    @classmethod
    def _validate_contract_version(cls, v: str) -> str:
        return validate_contract_version(v)

File: packages/relief_contracts/src/household_snapshot.ts
import { z } from "zod";
import { ContractVersion, Currency, AccountV1, ObligationV1, ProviderCapabilityV1, ConsumerConstitutionV1 } from "./shared";
import { FinancialEventV1 } from "./financial_event";

/**
 * HouseholdSnapshotV1 (Section 10)
 *
 * Owner: Plan Two
 * Version: 1.0.0
 * Purpose: the complete, self-contained input to every forecast provider. The
 *   model team must never query Plan Two's production database directly —
 *   Plan Two constructs and validates this snapshot instead.
 *
 * Required: all fields below (arrays may be empty, consumer_constitution may
 *   be an empty/default object if the consumer has not yet configured rules).
 * Optional: none at 1.0.0.
 *
 * Migration notes: none yet (first version).
 */
export const HouseholdSnapshotV1 = z.object({
  contract_version: ContractVersion,
  snapshot_id: z.string(),
  household_id: z.string(),
  generated_at: z.string().datetime({ offset: true }),
  timezone: z.string(),
  currency: Currency,
  accounts: z.array(AccountV1),
  obligations: z.array(ObligationV1),
  recent_events: z.array(FinancialEventV1),
  known_future_events: z.array(FinancialEventV1),
  consumer_constitution: ConsumerConstitutionV1,
  provider_capabilities: z.array(ProviderCapabilityV1),
});
export type HouseholdSnapshotV1 = z.infer<typeof HouseholdSnapshotV1>;
